File: Data Structures & Algorithms/subsets/submission-2.cs
public class Solution {
    public List<List<int>> Subsets(int[] nums) {
        List<List<int>> result = new List<List<int>>();
        List<int> subList = new List<int>();

        backtrack(result, subList, nums, 0);
        return result;
    }

    private void backtrack(List<List<int>> result, List<int> subList, int[] nums, int index){
        if(index >= nums.Length){
            result.Add(new List<int>(subList));
            return;
        }

        subList.Add(nums[index]);
        backtrack(result, subList, nums, index + 1);
        subList.RemoveAt(subList.Count - 1);
        backtrack(result, subList, nums, index + 1);

    }


}

File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> subList = new ArrayList<>();
        Arrays.sort(candidates);
        backtrack(candidates, result, subList, 0, target);

        return result;
    }

    private void backtrack(int[] nums, List<List<Integer>> result, List<Integer> subList,int index, int target){
        if(target == 0){
            result.add(new ArrayList<>(subList));
            return;
        }
        if(target < 0){
            return;
        }

        for(int i =index;i < nums.length; i++){
            if(i > index && nums[i] == nums[i - 1]){
                continue;
            }
            subList.add(nums[i]);
            backtrack(nums,result, subList, i + 1, target - nums[i]);
            subList.remove(subList.size() - 1);
        }


    }
}


